### PR TITLE
Manage the list of DDC nodes

### PR DESCRIPTION
**Pull request by doasync**
### User Story
**As a** SC client
**I want** to be able to register and list all DDC nodes
**So that** I can do it using SC functions

### New functions
- `add_ddc_node(id: String, url: String)`
- `get_all_ddc_nodes()`

### Tasks
[CBI-781](https://cerenetwork.atlassian.net/browse/CBI-781):
- [CBI-782](https://cerenetwork.atlassian.net/browse/CBI-782)
- [CBI-783](https://cerenetwork.atlassian.net/browse/CBI-783)

### CODE REVIEW

File: cere02/lib.rs [hunk 1]
```diff
@@ -36,6 +36,7 @@ mod ddc {
         // -- Metrics Reporting --
         pub metrics: StorageHashMap<MetricKey, MetricValue>,
         current_period_ms: u64,
+        ddn_nodes: Vec<DDNNode>
```
Comment (evgeny-s) [inline cere02/lib.rs:39]: Either change Name to DDCNode or just DDN.

Comment (doasync) [reply cere02/lib.rs:40]: I assumed DDN stands for "decentralized data network" ([source](https://blog.cere.network/cere-is-here-the-future-of-customer-data/)). Yeah, let's call them DDC nodes.

**Comment by doasync** [comment]
Do we need an event `DDCNodeAdded`?

File: cere02/lib.rs [hunk 2]
```diff
@@ -638,6 +640,35 @@ mod ddc {
     }
 
 
+    // ---- Admin: DDC nodes ----
+    #[derive(Default, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode, SpreadLayout, PackedLayout)]
+    #[cfg_attr(feature = "std", derive(Debug, scale_info::TypeInfo))]
+    pub struct DDCNode {
+        id: String,
```
Comment (naure) [inline cere02/lib.rs:647]: Please rename everywhere `id` to `p2p_id`, to not confuse with account ID. (refers to [libp2p](https://docs.libp2p.io/concepts/peer-id/))

File: cere02/lib.rs [hunk 3]
```diff
@@ -638,6 +640,35 @@ mod ddc {
...
+        #[ink(message)]
+        pub fn get_all_ddc_nodes(&self) -> Vec<DDCNode> {
+            self.ddc_nodes.values().cloned().collect()
+        }
+
+        /// Registers a DDC node
+        #[ink(message)]
+        pub fn add_ddc_node(&mut self, id: String, url: String) -> Result<()> {
+            let reporter = self.env().caller();
+            self.only_reporter(&reporter)?;
```
Comment (naure) [inline cere02/lib.rs:662]: This should be managed by the admin for now, not reporters.

**Comment by evgeny-s** [comment]
> Do we need an event DDCNodeAdded?
Yes, let's keep it

**Review by naure — CHANGES_REQUESTED** [review]
Nice work.

Let’s complete it right away with:
- Change to only_admin
- `remove_` and `is_` functions.
- Events for Add / Remove.

Basically the same as for reporters.
You may also move the code just after the "Admin: Reporters" section to show that it is similar.

File: cere02/lib.rs (hunk 2 above)
Comment (doasync) [reply cere02/lib.rs]: done

File: cere02/lib.rs (hunk 3 above)
Comment (doasync) [reply cere02/lib.rs]: done

**Comment by doasync** [comment]
@naure Thanks for your comments. I have updated the code:
- implemented emitting of add/remove events
- implemented `remove_` and `is_` functions
- set only_owner for add/remove
- renamed ddc node id to p2p_id
- moved ddc nodes code after the reporters
- refactored code

File: cere02/lib.rs [hunk 4]
```diff
@@ -488,6 +492,64 @@ mod ddc {
...
+    #[cfg_attr(feature = "std", derive(Debug, scale_info::TypeInfo))]
+    pub struct DDCNode {
+        p2p_id: String,
+        url: String,
+    }
+
+    #[ink(event)]
+    pub struct DDCNodeAdded {
+        #[ink(topic)]
+        p2p_id: String,
```
Comment (naure) [inline cere02/lib.rs:507]: There should be `url` too.

File: cere02/lib.rs [hunk 5]
```diff
@@ -949,6 +1010,11 @@ mod ddc {
             assert_eq!(contract.get_current_period_ms(), today_ms);
         }
 
+        fn decode_event(event: &ink_env::test::EmittedEvent) -> Event {
```
Comment (naure) [inline cere02/lib.rs:1013]: Good idea.

**Review by naure — APPROVED** [review]
Just one more request.

Great job.

File: cere02/lib.rs (hunk 4 above)
Comment (doasync) [reply cere02/lib.rs:507]: [done](https://github.com/Cerebellum-Network/cere-enterprise-smart-contracts/pull/20/files#diff-a99b45b57bfdc35f8b28ef154fa4418c459a459aa60dfe0f98d8a350a93fa067R508)

**Pull request merged**